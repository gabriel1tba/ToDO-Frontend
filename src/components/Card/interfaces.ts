export interface ICard {
  title: string;
  color?: 'info' | 'danger';
  description?: string;
  icon: React.ElementType;
}
