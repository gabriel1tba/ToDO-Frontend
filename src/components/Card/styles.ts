import styled, { css, DefaultTheme } from 'styled-components';

interface ContainerProps {
  color: 'info' | 'danger';
}

const colorModifiers = {
  info: (theme: DefaultTheme) => css`
    border: 1px solid ${theme.colors.gray[200]};
    color: ${theme.colors.gray[300]};
  `,
  danger: (theme: DefaultTheme) => css`
    border: 1px solid ${theme.colors.danger.light};
    color: ${theme.colors.danger.main};
  `,
};

export const Container = styled.div<ContainerProps>`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  margin: 1rem 0;
  border-radius: 8px;
  padding: 5rem 3rem;

  svg {
    font-size: 4rem;
    margin-bottom: 1rem;
  }

  strong {
    margin-bottom: 0.5rem;
    font-weight: bold;
  }

  ${({ theme, color }) => colorModifiers[color](theme)};
`;
