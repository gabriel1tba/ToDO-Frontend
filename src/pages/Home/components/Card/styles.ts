import styled, { css, DefaultTheme } from 'styled-components';

interface WrapperProps {
  color: 'info' | 'danger';
}

const colorModifiers = {
  info: ({ colors }: DefaultTheme) => css`
    border: 1px solid ${colors.gray[700]};
    color: ${colors.gray[300]};
  `,
  danger: ({ colors }: DefaultTheme) => css`
    border: 1px solid ${colors.danger.light};
    color: ${colors.danger.light};
  `,
};

export const Wrapper = styled.div<WrapperProps>`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  margin: 1rem 0;
  border-radius: 8px;
  padding: 5rem 3rem;

  svg {
    font-size: 4rem;
    margin-bottom: 1rem;
  }

  strong {
    margin-bottom: 0.5rem;
    font-weight: bold;
  }

  ${({ theme, color }) => colorModifiers[color](theme)};
`;
