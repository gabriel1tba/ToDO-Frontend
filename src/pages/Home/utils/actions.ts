import { TTodo } from 'services/TodoService/interfaces';

export enum Types {
  GET_TODOS = 'GET_TODOS',
  CREATE_TODO = 'CREATE_TODO',
  UPDATE_TODO = 'UPDATE_TODO',
  DELETE_TODO = 'DELETE_TODO',
}

interface GetTodos {
  type: Types.GET_TODOS;
  payload: TTodo[];
}

interface CreateTodo {
  type: Types.CREATE_TODO;
  payload: TTodo;
}

interface UpdateTodo {
  type: Types.UPDATE_TODO;
  payload: TTodo;
}

interface DeleteTodo {
  type: Types.DELETE_TODO;
  payload: string;
}

export type Action = GetTodos | CreateTodo | UpdateTodo | DeleteTodo;

export const getTodosAction = (todos: TTodo[]): GetTodos => ({
  type: Types.GET_TODOS,
  payload: todos,
});

export const createTodoAction = (todo: TTodo): CreateTodo => ({
  type: Types.CREATE_TODO,
  payload: todo,
});

export const updateTodoAction = (todo: TTodo): UpdateTodo => ({
  type: Types.UPDATE_TODO,
  payload: todo,
});

export const deleteTodoAction = (id: string): DeleteTodo => ({
  type: Types.DELETE_TODO,
  payload: id,
});
