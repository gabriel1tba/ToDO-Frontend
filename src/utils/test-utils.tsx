import { render, RenderOptions } from '@testing-library/react';
import { TodoProvider } from 'context/todos';

interface IWrapperProvider {
  children: React.ReactNode;
}

const WrapperProvider = ({ children }: IWrapperProvider) => (
  <TodoProvider>{children}</TodoProvider>
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(<WrapperProvider>{ui}</WrapperProvider>, { ...options });

export * from '@testing-library/react';
export { customRender as render };
