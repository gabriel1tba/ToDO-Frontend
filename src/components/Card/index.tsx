import * as S from './styles';

export interface ICard {
  title: string;
  color?: 'info' | 'danger';
  icon: React.ElementType;
  children?: React.ReactNode;
}

const Card = ({ title, icon: Icon, color = 'info', children }: ICard) => (
  <S.Wrapper color={color}>
    <Icon />
    <strong>{title}</strong>

    {children && children}
  </S.Wrapper>
);

export default Card;
