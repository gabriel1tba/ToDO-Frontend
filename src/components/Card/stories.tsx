import { Story, Meta } from '@storybook/react';
import { CgNotes } from 'react-icons/cg';

import Card from '.';
import { ICard } from './interfaces';

export default {
  title: 'Card',
  component: Card,
  argTypes: {
    title: {
      type: 'string',
    },
    description: {
      type: 'string',
    },
    color: {
      control: {
        type: 'select',
        options: ['info', 'danger'],
      },
    },
  },
} as Meta;

export const Default: Story<ICard> = (args) => (
  <div style={{ maxWidth: '30rem', margin: '0 auto' }}>
    <Card {...args} />
  </div>
);

Default.args = {
  title: 'Você ainda não tem tarefas cadastradas',
  description: 'Crie tarefas clicando no botão Nova Tarefa',
  icon: CgNotes,
};

export const Danger: Story<ICard> = (args) => (
  <div style={{ maxWidth: '30rem', margin: '0 auto' }}>
    <Card {...args} />
  </div>
);

Danger.args = {
  color: 'danger',
  title: 'Você ainda não tem tarefas cadastradas',
  icon: CgNotes,
};
