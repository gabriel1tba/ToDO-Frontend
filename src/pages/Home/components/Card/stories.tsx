import { Story, Meta } from '@storybook/react';
import { CgNotes } from 'react-icons/cg';

import Card, { ICard } from '.';

export default {
  title: 'Card',
  component: Card,
  argTypes: {
    title: {
      type: 'string',
    },
    description: {
      type: 'string',
    },
    color: {
      control: {
        type: 'select',
        options: ['info', 'danger'],
      },
    },
  },
} as Meta;

export const Default: Story<ICard> = (args) => (
  <div style={{ maxWidth: '30rem', margin: '0 auto' }}>
    <Card {...args} />
  </div>
);

Default.args = {
  title: 'Você ainda não tem tarefas cadastradas',
  icon: CgNotes,
};

export const Danger: Story<ICard> = (args) => (
  <div style={{ maxWidth: '30rem', margin: '0 auto' }}>
    <Card {...args} />
  </div>
);

Danger.args = {
  color: 'danger',
  title: 'Você ainda não tem tarefas cadastradas',
  icon: CgNotes,
};
