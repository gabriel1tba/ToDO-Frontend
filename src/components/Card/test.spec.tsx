import { FaUser } from 'react-icons/fa';
import { render, screen } from 'utils/test-utils';

import Card from '.';

describe('<Card />', () => {
  beforeEach(() => {
    render(<div id="portal-modal-root" />);
  });

  it('should render the Card correctly', () => {
    const { container } = render(
      <Card icon={FaUser} title="Card Title" description="Card Description" />
    );

    expect(screen.getByText(/card title/i)).toBeInTheDocument();
    expect(screen.getByText(/card description/i)).toBeInTheDocument();

    expect(screen.getByText(/card title/i).parentElement).toHaveStyle({
      color: '#616161',
      border: '1px solid #BCBCBC',
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the Card variant Danger correctly', () => {
    render(
      <Card
        icon={FaUser}
        title="Card Title"
        description="Card Description"
        color="danger"
      />
    );

    expect(screen.getByText(/card title/i).parentElement).toHaveStyle({
      color: '#FF1744',
      border: '1px solid #FF5252',
    });
  });
});
