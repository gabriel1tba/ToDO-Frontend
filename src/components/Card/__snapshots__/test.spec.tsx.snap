// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render the Card correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  margin: 1rem 0;
  border-radius: 8px;
  padding: 5rem 3rem;
  border: 1px solid #BCBCBC;
  color: #616161;
}

.c0 svg {
  font-size: 4rem;
  margin-bottom: 1rem;
}

.c0 strong {
  margin-bottom: 0.5rem;
  font-weight: bold;
}

<div
  class="c0"
  color="info"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 448 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
    />
  </svg>
  <strong>
    Card Title
  </strong>
  <p>
    Card Description
  </p>
</div>
`;
